keps/sig-network/20190603-EndpointSlice-API.md

```diff
@@ -0,0 +1,346 @@
+| # of writes              | O(P)                        | O(P)                            | O(P)                         |
+|                          | 20                          | 20                              | 20                           |
+| Size of API object       | O(P)                        | O(B)                            | O(1)                         |
+|                          | ~1KB                        | ~1KB                            | ~1KB                         |
+| # of watchers per object | O(N)                        | O(N)                            | O(N)                         |
+|                          | 10                          | 10                              | 10                           |
+| # of total watch event   | O(NP)                       | O(NP)                           | O(NP)                        |
+|                          | 10 * 20                     | 10 * 20                         | 10 * 20                      |
+| Total Bytes Transmitted  | O(P^2N)                     | O(NPB)                          | O(NP)                        |
+|                          | ~1KB * 10 * 20 = 200KB      | ~1KB * 10 * 20 = 200KB          | ~1KB * 10 * 20 = 200KB       |
+
+
+## Implementation
+
+### Requirements
+
+- Persistence (Minimal Churn of Endpoints)
+
+Upon service endpoint changes, the # of object writes and disruption to ongoing connections should be minimized. 
+
+- Handling Restarts & Failures
+
+The producer/consumer of EndpointSlice must be able to handle restarts and recreate state from scratch with minimized change to existing state.
+ 
+
+### EndpointSlice Controller
+
+A new EndpointSlice Controller will be added to `kube-controller-manager`. It will manage the lifecycle EndpointSlice instances derived from  
+```
+Watch: Service, Pod ==> Manage: EndpointSlice
+```
+
+#### Workflows
+On Service Create/Update/Delete:
+- `syncService(svc)`
+
+On Pod Create/Update/Delete: 
+- Reverse lookup relevant services
+- For each relevant service, 
+  - `syncService(svc)`
+  
+
+`syncService(svc)`:
+- Look up selected backend pods
+- For each pod 
+  - If pod is already added into EndpointSlice and Status is correct
+    - Skip
+  - If Pod needs to be added:
+    - If  all EndpointSlice objects has reached the MaxEndpointThreshold. Create a new EndpointSlice object.
+    - Find the EndpointSlice with the lowest number of endpoint, and add it into it. 
+  - If Pod needs to be updated:
+    - Update the corresponding EndpointSlice
+  - If Pod needs to be removed:
+    - Remove the endpoint from EndpointSlice. 
+    - If the # of endpoints in EndpointSlice is less than ¼ of threshold.
+      - Try to pack the endpoints into smaller number of EndpointSlice objects.
+
+### Kube-Proxy
+
+Kube-proxy will be modified to consume EndpointSlice instances besides Endpoints resource. A flag will be added to kube-proxy to toggle the mode. 
```

Comment:
Is there a way to autodetect this? eg start with Endpoints but the moment we see an EndpointSlice switch to watching those? Seems like a shame to add another flag.

Reply:
How about this?
The controller creates a config map with specific name.
kube-proxy tries to get the config map on start up. If the config map exists, then use EndpointSlice, if not fall back to Endpoints.

Reply:
If all goes well, I think EndpointSlice and TopologyAwareServiceRouting goes Alpha in 1.16 together, which EndpointSlice implement controller only, and TopologyAwareServiceRouting watch node for topology infomation. EndpointSlice and TopologyAwareServiceRouting goes Beta in 1.17 together, which  kube-proxy will consume  EndpointSlice, TopologyAwareServiceRouting also rely on this, so what about consuming EndpointSlice when TopologyAwareServiceRouting is enabled.